src/Microsoft.AspNet.OData.Shared/EdmEntityObject.cs

```diff
@@ -42,6 +46,62 @@ public EdmEntityObject(IEdmEntityTypeReference edmType)
         public EdmEntityObject(IEdmEntityType edmType, bool isNullable)
             : base(edmType, isNullable)
         {
+            PersistentInstanceAnnotationsContainer = new ODataInstanceAnnotationContainer();
+            TransientInstanceAnnotationContainer = new ODataInstanceAnnotationContainer();
+        }
+
+        /// <summary>
+        /// Instance Annotation container to hold Transient Annotations.
+        /// </summary>
+        internal IODataInstanceAnnotationContainer TransientInstanceAnnotationContainer { get; set; }
+
+        /// <summary>
+        /// Instance Annotation container to hold Persistent Annotations.
+        /// </summary>
+        public IODataInstanceAnnotationContainer PersistentInstanceAnnotationsContainer { get; set; }
+
+        /// <summary>
+        /// Container to hold ODataId.
+        /// </summary>
+        public IODataIdContainer ODataIdContainer { get; set; }
+
+        /// <summary>
+        /// OData Path for the Item.
+        /// </summary>
+        internal ODataPath ODataPath { get; set; }
+
+        /// <summary>
+        /// DeltaKind as Entry
+        /// </summary>
+        public virtual EdmDeltaEntityKind DeltaKind { get { return EdmDeltaEntityKind.Entry; } }
+
+        /// <summary>
+        /// Method to Add Data Modification Exception.
+        /// </summary>
+        public void AddDataException(DataModificationExceptionType dataModificationException)
+        {
+            if (TransientInstanceAnnotationContainer == null)
+            {
+                throw Error.ArgumentNull(nameof(TransientInstanceAnnotationContainer));
```

Comment:
I don't think it's right to throw an `ArgumentNullException` because that suggests that a null was passed in as argument